JP-2807: Update ATOCA algorithm

tapastro opened the pull request:
<!-- If this PR closes a JIRA ticket, make sure the title starts with the JIRA issue number, 
for example JP-1234: <Fix a bug> -->
Resolves [JP-2807](https://jira.stsci.edu/browse/JP-2807)

<!-- If this PR closes a GitHub issue, reference it here by its number -->
Closes #

<!-- describe the changes comprising this PR here -->
This PR updates the ATOCA algorithm from the initial merge to the current development branch, increasing stability and reducing runtime. The modeling is now off by default, and some documentation is required to cover the new step arguments.

**Checklist**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [x] updated or added relevant tests
- [x] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)

tapastro commented:
Converting from draft was done too quickly - this isn't quite ready yet.

stscieisenhamer commented on jwst/extract_1d/soss_extract/atoca.py:
```diff
@@ -72,6 +76,13 @@ def __init__(self, wave_map, trace_profile, throughput, kernels,
+            Boolean Mask of the detector pixels to mask for every extraction. 
+            Should not be related to a specific order (if so, use `mask_trace_profile` instead).
+        mask_trace_profile: (N_ord, N, M) list or array of 2-D arrays[bool], optional
+            A list or array of the pixel that need to be used for extraction,
```
spelling: "pixel" -> "pixels"

stscieisenhamer commented on jwst/extract_1d/soss_extract/atoca.py:
```diff
@@ -129,6 +142,13 @@ def __init__(self, wave_map, trace_profile, throughput, kernels,
+        # Set the mask based on trace profiles and save
+        if mask_trace_profile is None:
+            # No mask (False everywhere)
+            log.warning('`mask_trace_profile` was not given. Not optimal.')
```
a rather terse warning. Is there more to be said or a documentation reference?

tapastro replied:
@AntoineDarveau thoughts? 🙂

AntoineDarveau replied:
I agree 😅. It is not optimal because the whole detector will be model. It is more optimal (in memory and time) to model only the region of interest around the traces. So the message could say:
'`mask_trace_profile` was not given. All detector's pixels will be modeled.'

stscieisenhamer commented on jwst/extract_1d/soss_extract/atoca.py:
```diff
@@ -1148,6 +1091,25 @@ def best_tikho_factor(self, tests=None, fit_mode='all', mode_kwargs=None):
             tests = self.tikho_tests
 
+        # TODO Find a way to identify when the solution becomes unstable
+        # and do nnot use these in the search for the best tikhonov factor.
```
spelling: "nnot" -> "not"

stscieisenhamer commented on jwst/extract_1d/soss_extract/atoca.py:
```diff
@@ -1148,6 +1091,25 @@ def best_tikho_factor(self, tests=None, fit_mode='all', mode_kwargs=None):
         if tests is None:
             tests = self.tikho_tests
 
+        # TODO Find a way to identify when the solution becomes unstable
```
suggestion: make a jp issue about this.

stscieisenhamer commented on jwst/extract_1d/soss_extract/atoca.py:
```diff
@@ -1198,6 +1160,8 @@ def best_tikho_factor(self, tests=None, fit_mode='all', mode_kwargs=None):
         # Get the factor of the chosen mode
         best_fac = results[best_mode]
 
+        log.debug(f'mode chosen to find regularization factor is {best_mode}')
```
"mode" -> "Mode"

stscieisenhamer commented on jwst/extract_1d/soss_extract/atoca.py:
```diff
@@ -1506,6 +1470,9 @@ class ExtractionEngine(_BaseOverlap):
     This version models the pixels of the detector using an oversampled trapezoidal integration.
     """
 
+    # The desired data-type for computations, e.g., 'float32'. 'float64' is recommended.
```
did this not already happen in this module? maybe unrelated to the first instance, but consider a refactor to avoid this being specified twice.

hbushouse replied:
Yes, there's already an instance of this up above in atoca.py (same module)

stscieisenhamer commented on jwst/extract_1d/soss_extract/atoca_utils.py:
```diff
@@ -10,7 +10,7 @@
 from scipy.sparse.linalg import spsolve
 from scipy.interpolate import interp1d, RectBivariateSpline, Akima1DInterpolator
-from scipy.optimize import minimize_scalar, newton
+from scipy.optimize import minimize_scalar, brentq
```
Though normally I ignore, there are a large number of annotations about test coverage here. I seem to have read that this is "still in the works". Guess is Howard ok with this?

stscieisenhamer replied:
Re-iterating the concern of apparent lack of test coverage.

hbushouse replied:
Agree that we at least need more regtests of this extraction applied with various param settings, etc.

stscieisenhamer commented on jwst/extract_1d/soss_extract/soss_extract.py:
```diff
@@ -160,19 +161,21 @@ def estim_flux_first_order(scidata_bkg, scierr, scimask, ref_files, threshold):
+        Mask determining the aperture used for extraction. Set to False where the pixel should be extracted.
+    threshold : float, optional:
+        The pixels with an aperture[order 2] > `threshold` are considered contaminated
+        and will be masked. Default is 1e-4.
```
documentation and code are not in sync: 1e-2 vs 1e-4.

stscieisenhamer commented on jwst/extract_1d/soss_extract/soss_solver.py:
```diff
@@ -129,7 +129,6 @@ def _chi_squared(transform, xref_o1, yref_o1, xref_o2, yref_o2,
     chisq : float
```
Same comment as earlier about the extensive non-test-coverage.

tapastro commented:
Updated regression tests with coverage for new ATOCA flags - run here: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/470/

tapastro commented on CHANGES.rst:
```diff
@@ -48,6 +48,8 @@ extract_1d
 - Fix IFU spectral extraction code to not fail on NaN fill values
   that now populate empty regions of the data cubes. [#7337]
 
+- Update ATOCA algorithm to development version [#6945]
```
Updated in latest commits.

tapastro commented:
Updated tests and responded to review comments. Running another regtest set: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/475/

tapastro commented on jwst/extract_1d/soss_extract/soss_boxextract.py:
```diff
@@ -98,17 +98,25 @@ def box_extract(scidata, scierr, scimask, box_weights, cols=None):
 
+    # Set all pixels values outside of extraction region to Nan
+    # so it will be correctly handle by np.nansum.
+    data = np.where(extract_region, data, np.nan)
+    error = np.where(extract_region, error, np.nan)
```
@AntoineDarveau found a bug in the boxextract code which I think exists in your latest branch as well (not to say that you wrote it - quite possible it filtered in from merging some of my commits!) Just wanted to warn you, as it prevents any data output.

tapastro commented:
@hbushouse This could use a review.

hbushouse commented on docs/jwst/extract_1d/arguments.rst:
```diff
@@ -89,38 +89,68 @@ The ``extract_1d`` step has the following step-specific arguments.
+
+``--soss_wave_grid_in``
+  This is a NIRISS-SOSS algorithm-specific parameter; filename or SossWaveGridModel
+  containing the wavelength grid used by ATOCA to model each pixel valid pixel of the
```
typo: "... model each pixel valid pixel ..."

hbushouse commented on jwst/extract_1d/soss_extract/atoca.py:
```diff
@@ -70,8 +74,13 @@ def __init__(self, wave_map, trace_profile, throughput, kernels,
-        orders: list, optional
+            Boolean Mask of the detector pixels to mask for every extraction.
+            Should not be related to a specific order (if so, use `mask_trace_profile` instead).
+        mask_trace_profile: (N_ord, N, M) list or array of 2-D arrays[bool], optional
```
minor nit: should have whitespace between end of argument name and ":" (applies to "orders" too)

hbushouse commented on jwst/extract_1d/soss_extract/atoca.py:
```diff
@@ -1538,7 +1513,12 @@ def __init__(self, wave_map, trace_profile, *args, **kwargs):
-            Boolean mask of the bad pixels on the detector.
+            Boolean Mask of the detector pixels to mask for every extraction.
+            Should not be related to a specific order (if so, use `mask_trace_profile` instead).
+        mask_trace_profile: (N_ord, N, M) list or array of 2-D arrays[bool], optional
```
whitespace before ":"

hbushouse commented on jwst/extract_1d/soss_extract/atoca.py:
```diff
@@ -1538,7 +1513,12 @@ def __init__(self, wave_map, trace_profile, *args, **kwargs):
+            A list or array of the pixel that need to be used for extraction,
+            for each order on the detector. It has to have the same (N_ord, N, M) as `trace_profile`.
+            If not given, `threshold` will be applied on spatial profiles to define the masks.
         orders: list, optional
```
whitespace before ":"

hbushouse commented on jwst/extract_1d/soss_extract/atoca_utils.py:
```diff
@@ -618,6 +618,157 @@ def get_soss_grid(wave_maps, trace_profiles, wave_min=0.55, wave_max=3.0, n_os=N
+
+    Parameters
+    ----------
+    all_grid: list[array]
```
need whitespace before ":" for all args listed here

hbushouse reviewed (commented):
Too much code to do a complete review!  Just minor comments.

tapastro commented:
Cleaned up the review comments, apart from the testing coverage - I added a regression test that should add significant coverage, but Codecov will only show once it enters our nightly runs. I am happy to add more tests if it shows some gaps.

hbushouse reviewed (approved):
Works for me.

hbushouse commented:
Is there a recent regtest run for this?

tapastro commented:
476 was the last run, from which only a merge to master and docstring updates have been added since. I can run another, though.

tapastro commented:
Started another: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/486/

hbushouse commented:
Do these look like expected differences? https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/486/testReport/jwst.regtest/test_niriss_soss/

tapastro commented:
> Do these look like expected differences? https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/486/testReport/jwst.regtest/test_niriss_soss/

The `test_niriss_soss_extras` truth files were generated on my machine, so platform differences are expected - they should be fine. The `test_niriss_soss_stage3` results should also change, as it now uses ATOCA instead of box extraction.

The stage2 difference worried me a bit, but I inspected each output spectrum by eye and there is no appreciable difference. There are also some changes to SOSS-specific extraction parameters SOSSDLTX, SOSSDLTY and SOSSROTN that could likely cause some slight deviation in output.

So in all, I think the changes are expected.

hbushouse commented:
Looks good except for the usual suspects in the CI tests, so I'm merging.
